### ISSUE TITLE
BO - Order page - when adding a product which has an ecotax to an old order, the base price is NOK

### ISSUE THREAD

**Issue by khouloudbelguith** [PrestaShop/PrestaShop#22029]
#### Describe the bug

The ecotax is enabled, In the BO > Orders > old order details page > when we try to add a product which has an ecotax, the base is NOK, so the total is NOK, the invoice is NOK

The Edit of a product with Ecotax is NOK too.

#### Steps to Reproduce

0. Go to BO > International > Taxes > Enable "Use ecotax"
![image](https://user-images.githubusercontent.com/16067358/99795945-ef671e00-2b2c-11eb-82a1-a362773d2d90.png)
1. Create an order from the FO with any product
2. Go to BO > Catalog > Products page > Create a new product
2.1 Tax included = 12€
2.2 Tax rule = 20%
2.3 Ecotax (tax incl.) = 5€ 
![image](https://user-images.githubusercontent.com/16067358/99796984-b3cd5380-2b2e-11eb-9c57-372fe97139a8.png)
3. Go to BO > Orders > last order created
4. Add a product
5. Search the last product created
![image](https://user-images.githubusercontent.com/16067358/99796700-399ccf00-2b2e-11eb-83fd-1268647d1397.png)
6. Click add
7. see the error
![image](https://user-images.githubusercontent.com/16067358/99796739-4b7e7200-2b2e-11eb-93d3-9769b0201564.png)
8. change the status to payment accepted
9. Download invoice
10. see error
![image](https://user-images.githubusercontent.com/16067358/99796842-7799f300-2b2e-11eb-9f29-27350ca8cfc9.png)

PS1768 => ok: https://drive.google.com/file/d/12s52BuZBTIUykgD6j4mRTv-5RhD_xtPA/view?usp=sharing
177x => NOK: https://drive.google.com/file/d/1X1F-PonwH9O1CBsm_WfCsUsW8NsXyLux/view?usp=sharing

#### Additional information

* PrestaShop version: 177x
* PHP version: 7.2

-----
## Expected behavior

As the ecotax is included in the price filled by the merchant, the unit prices should be displayed ecotax included
Specs: https://github.com/PrestaShop/prestashop-specs/blob/master/broader-topics/ecotax.md
Note: For the ecotax VAT rate, take into account the one defined in International > Taxes

## Technical viewpoint (@jolelievre)

This comment should help you regarding the technical solution that needs to be implemented https://github.com/PrestaShop/PrestaShop/issues/22029#issuecomment-767658294 If you have any question don't hesitate to ask me

## Status

Waiting for https://github.com/PrestaShop/PrestaShop/pull/22228 to be merged

**Comment by matks** [comment]
Hu @khouloudbelguith in your invoice I see that there is 2 lines
- 1st line: this is the product you added
- 2nd line for "demo_8"

It seems the 2nd line is wrong too ? 🤔  I see the tax included is equal to tax excluded, although tax rate is supposed to be 20%/

**Comment by Robin-Fischer-PS** [comment]
@matks The second line has not "tax incl." column, it's only Unit Price and Total (both Tax excl.).

**Comment by khouloudbelguith** [comment]
Hi @matks,

The demo_8 from the BO
![image](https://user-images.githubusercontent.com/16067358/99807206-732a0600-2b3f-11eb-9f15-4a085e262aa1.png)

Thanks!

**Comment by matks** [comment]
> @matks The second line has not "tax incl." column, it's only Unit Price and Total (both Tax excl.).

I am asking about the _invoice_ 😄 . This:

<img src="https://user-images.githubusercontent.com/16067358/99796842-7799f300-2b2e-11eb-9f29-27350ca8cfc9.png">

**Comment by matks** [comment]
@khouloudbelguith my bad I am really tired, there is no issue

**Comment by Robin-Fischer-PS** [comment]
This issue also happens when editing a product with Ecotax (Prices are displayed OK during Edit, but the Ecotax is added once the Edit validated, it shouldn't).

**Comment by matks** [comment]
I thought that `$product->price` contained the ecotax but no. So when Khouloud does this
```
2.1 Tax included = 12€
2.2 Tax rule = 20%
2.3 Ecotax (tax incl.) = 5€
```

What happens in database is
- product->price = 5.83 (without taxes)
- product->ecotax = 5
and consequently `product->unit_price` is the sum of these two so we have (outside of database, this is computed)
- product->unit_price_without_taxes = 10.83 (unit = with ecotax included)
- product->unit_price_with_taxes = 12

### What happens with the bug

When we open the Order and want to add a product, we search for the product. When the product search is done, it returns results. **The prices returned for this result were product->unit_price_without_taxes and product->unit_price_with_taxes**.

However when we click "add to order" button, these 2 properties are used to fill the database, meaning
- product->unit_price_without_taxes is inserted as product->price
And then the code sees that "oh! this is a product that has ecotax" and add it.

### Small solution

Here is a beginning for a fix https://github.com/matks/PrestaShop/commit/f4148ce38c9239d381a9a3e6ca05c7705e77d8a4
I simply modify the "search products" logic to return prices WITHOUT ecotax. So
```
The prices returned for this result were product->unit_price_without_taxes and product->unit_price_with_taxes.
```
becomes
```
The prices returned for this result were product->price and product->unit_price_with_taxes.
```

The problem is that the merchant sees this
![Capture d’écran 2020-11-20 à 17 38 55](https://user-images.githubusercontent.com/3830050/99825457-5ac5e580-2b57-11eb-8637-45b6a4e62a7c.png)

And when he clicks "add to order" he sees that
![Capture d’écran 2020-11-20 à 17 39 02](https://user-images.githubusercontent.com/3830050/99825477-5f8a9980-2b57-11eb-8d9a-5c8aaa5979a5.png)

Because
- he has seen the search results data, which ignores ecotax
- then when product has been added to order, he sees "prices with ecotax included"

So ... I guess we need to _highlight the fact that this is "without ecotax"? Then we need _to add new data into the response of the "search a product" logic. And display them.

**Comment by matks** [comment]
**After very heavy discussion** with @eternoendless @marionf @jolelievre @Progi1984 and many more we found many more questions and issues about the use of ecotax in this BO page and in PrestaShop in general, too much to be handled in 1.7.7 at this stage of the release. We will even need deep exploration of the topic to correctly handle it.

**Comment by MatShir** [comment]
One solution:
As a merchant, I can't edit the price below the eco-tax

**Comment by jolelievre** [comment]
I think the error is related to the base price saved for the product which considers the input is the base price and it's the same as product price

But actually `product_price = base_price + eco_tax` so when we save the input as base price the eco tax should be substracted Besides when we add or edit a price in the Order when price with taxes doesn't match the price without taxes we recompute the base price based on applied tax rate so that they are coherent with each other

There was a huge refacto regarding SpecificPrice in the order management, so first check that the error is still present If the problem is still present then my guess would be that it should be fixed in this class https://github.com/PrestaShop/PrestaShop/blob/1.7.7.x/src/Adapter/Order/OrderDetailUpdater.php

And regarding the price recomputing more precisely it happens here:
https://github.com/PrestaShop/PrestaShop/blob/d5bd4819cbea60e6aa383ecab983779595181de9/src/Adapter/Order/OrderDetailUpdater.php#L367

If the ecotax substraction is needed my best guess is here, but the `ecotax` field of OrderDetail hasn't been handled at all! So it should probably be managed inside this `OrderDetailUpdater` service (and not only in the `getPrecisePriceTaxExcluded` function) And in case the merchant tries to set a price that is inferior to the amount of the eco tax an exception should be thrown to forbid such modification

**Comment by matks** [comment]
@MatShir Is this issue supposed to be fixed for 1.7.7.2 ? It's marked as `critical`.

**Comment by PierreRambaud** [comment]
@matks if we can fix it for the 1.7.7.2 yes, otherwise it will wait for the next patch release :sweat_smile:

**Comment by MatShir** [comment]
Yes, it is on the ToDo column, so I should be picked ASAP

**Comment by marionf** [comment]
Waiting for https://github.com/PrestaShop/PrestaShop/pull/22228 to be merged

**Comment by matks** [comment]
@MatShir @marionf Just discovered https://github.com/PrestaShop/PrestaShop/issues/10030 😆

**Comment by marionf** [comment]
@matks, yes it's another issue about the ecotax

**Comment by khouloudbelguith** [comment]
Hi!

When testing this [PR](https://github.com/PrestaShop/PrestaShop/pull/23492), I created two issues:
https://github.com/PrestaShop/PrestaShop/issues/24340
https://github.com/PrestaShop/PrestaShop/issues/24373

Thank you!

**Comment by Progi1984** [comment]
Fixed by #23492
